pkg/controller/direct/privilegedaccessmanager/entitlement_controller.go

```diff
@@ -0,0 +1,442 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "privilegedaccessmanager-controller"
+	serviceDomain = "//privilegedaccessmanager.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.PrivilegedAccessManagerEntitlementGVK, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &entitlementModel{config: *config}, nil
+}
+
+var _ directbase.Model = &entitlementModel{}
+
+type entitlementModel struct {
+	config config.ControllerConfig
+}
+
+func (m *entitlementModel) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building PrivilegedAccessManager client for Entitlement: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *entitlementModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+
+	obj := &krm.PrivilegedAccessManagerEntitlement{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	container, err := oneOfContainer(ctx, reader, obj,
+		obj.Spec.ProjectRef,
+		obj.Spec.FolderRef,
+		obj.Spec.OrganizationRef)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving 'obj.Spec.ProjectRef', "+
+			"'obj.Spec.FolderRef' and 'obj.Spec.OrganizationRef': %w", err)
+	}
+
+	// Get location
+	location := *obj.Spec.Location
+
+	var id *PrivilegedAccessManagerEntitlementIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(container, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.Container != container {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has parent container changed, expected %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Container, container)
+		}
+		if id.Parent.Location != location {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Location, location)
+		}
+		if id.Entitlement != resourceID {
+			return nil, fmt.Errorf("PrivilegedAccessManagerEntitlement %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Entitlement, resourceID)
+		}
+	}
+
+	container, err = oneOfContainer(ctx, reader, obj,
```

Comment:
Nit: it might be clearer _not_ to reuse the `container` variable name

Reply:
Will make the change in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2659.